internal/pkg/workspace/workspace.go

```diff
@@ -236,31 +241,55 @@ func (ws *Workspace) ListAddonFiles(appName string) ([]string, error) {
 	return addonsFileNames, nil
 }
 
-// ReadAddonFiles reads all addon files.
-func (ws *Workspace) ReadAddonFiles(appName string) (*ReadAddonFilesOutput, error) {
-	addonFiles, err := ws.ListAddonFiles(appName)
+// ReadAddonFiles reads addon files for an application.
+func (ws *Workspace) ReadAddonFiles(appName string) (*AddonFiles, error) {
+	addonFileNames, err := ws.ListAddonFiles(appName)
 	if err != nil {
-		return nil, fmt.Errorf("list addon files: %w", err)
+		return nil, err
 	}
-	var resources, params, outputs string
-	for _, fileName := range addonFiles {
+	addonFiles := map[string]string{paramsFile: "", outputsFile: "", resourcesFiles: ""}
+	for _, fileName := range addonFileNames {
 		content, err := ws.read(appName, addonsDirName, fileName)
+		// Trim spaces at the end of the file in case unexpected "\n" to make the template nicer.
+		trimmedContent := strings.TrimSpace(string(content))
 		if err != nil {
 			return nil, fmt.Errorf("read addon file %s: %w", fileName, err)
 		}
 		switch fileName {
-		case paramsFileName:
-			params = string(content)
-		case outputsFileName:
-			outputs = string(content)
+		case paramsFile:
+			addonFiles[paramsFile] = trimmedContent
+		case outputsFile:
+			addonFiles[outputsFile] = trimmedContent
 		default:
-			resources += string(content)
+			addonFiles[resourcesFiles] += trimmedContent + "\n"
+		}
+	}
+	var missingFiles []string
+	resourceMissing := false
+	for fileName, content := range addonFiles {
+		if content != "" {
+			continue
+		}
+		if fileName == resourcesFiles {
+			resourceMissing = true
+		} else {
+			missingFiles = append(missingFiles, fileName)
 		}
 	}
-	return &ReadAddonFilesOutput{
-		Parameters: strings.Split(params, "\n"),
-		Resources:  strings.Split(resources, "\n"),
-		Outputs:    strings.Split(outputs, "\n"),
+	sort.Strings(missingFiles)
+	if len(missingFiles) != 0 {
+		if resourceMissing {
+			return nil, fmt.Errorf(`addons directory missing file(s): %s, and at least one resource YAML file such as "s3-bucket.yml"`, strings.Join(missingFiles, ", "))
+		}
+		return nil, fmt.Errorf(`addons directory missing file(s): %s`, strings.Join(missingFiles, ", "))
+	}
+	if resourceMissing {
+		return nil, fmt.Errorf(`addons directory missing file(s): at least one resource YAML file such as "s3-bucket.yml"`)
+	}
+	return &AddonFiles{
+		Parameters: strings.Split(strings.TrimSpace(addonFiles[paramsFile]), "\n"),
+		Resources:  strings.Split(strings.TrimSpace(addonFiles[resourcesFiles]), "\n"),
+		Outputs:    strings.Split(strings.TrimSpace(addonFiles[outputsFile]), "\n"),
```

Comment:
I think we can remove the `resourceMissing` and possibly simplify some of this logic, let me know what you think of this:
```go
   // same until the switch statement
   f := &AddonFiles{
        Parameters: strings.Split(strings.TrimSpace(addonFiles[paramsFile]), "\n"),
        Resources:  strings.Split(strings.TrimSpace(addonFiles[resourcesFiles]), "\n"),
        Outputs:    strings.Split(strings.TrimSpace(addonFiles[outputsFile]), "\n"),
   }
   return f, validateNoMissingFiles(f)
}

func validateNoMissingFiles(f *AddonFiles) error {
   var missingFiles []string
   if f.Params == nil {
       missingFiles = append(missingFiles, paramsFile)
   }
   if f.Outputs == nil {
       missingFiles = append(missingFiles, outputsFile)
   }
   if f.Resources == nil {
      missingFiles = append(missingFiles, `at least one resource YAML file such as "s3-bucket.yml"`)
   }

   if missingFiles != nil {
        return fmt.Errorf("addons directory has missing file(s): %s", strings.Join(missingFiles, ","))
   }
   return nil
}
```